File: jskat-swing-gui/src/main/java/org/jskat/gui/swing/table/GameAnnouncePanel.java
package org.jskat.gui.swing.table;

import org.jskat.control.JSkatEventBus;
import org.jskat.control.event.skatgame.InvalidNumberOfCardsInDiscardedSkatEvent;
import org.jskat.control.gui.action.JSkatAction;
import org.jskat.data.GameAnnouncement;
import org.jskat.data.GameContract;
import org.jskat.data.JSkatOptions;
import org.jskat.gui.swing.LayoutFactory;
import org.jskat.util.CardList;
import org.jskat.util.GameType;
import org.jskat.util.JSkatResourceBundle;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.swing.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.ItemEvent;
import java.awt.event.ItemListener;

/**
 * Holds widgets for announcing a game
 */
class GameAnnouncePanel extends JPanel {


    private static final Logger log = LoggerFactory.getLogger(GameAnnouncePanel.class);

    JSkatResourceBundle strings;
    JSkatOptions options;

    JRadioButton grandRadioButton;
    JRadioButton clubsRadioButton;
    JRadioButton spadesRadioButton;
    JRadioButton heartsRadioButton;
    JRadioButton diamondsRadioButton;
    JRadioButton nullRadioButton;
    ButtonGroup gameTypeButtonGroup;
    JCheckBox handBox = null;
    JCheckBox ouvertBox = null;
    JCheckBox schneiderBox = null;
    JCheckBox schwarzBox = null;

    DiscardPanel discardPanel;
    JSkatUserPanel userPanel;

    boolean userPickedUpSkat = false;

    /**
     * Constructor
     *
     * @param actions      Action map
     * @param userPanel    User panel
     * @param discardPanel Discard panel
     */
    GameAnnouncePanel(final ActionMap actions, final JSkatUserPanel userPanel,
                      final DiscardPanel discardPanel) {

        strings = JSkatResourceBundle.INSTANCE;
        this.userPanel = userPanel;
        this.discardPanel = discardPanel;

        initPanel(actions);
    }

    private void initPanel(final ActionMap actions) {

        setLayout(LayoutFactory.getMigLayout("fill"));

        final JPanel panel = new JPanel(LayoutFactory.getMigLayout("fill"));

        handBox = new JCheckBox(strings.getString("hand"));
        handBox.setEnabled(false);
        ouvertBox = createOuvertBox();
        schneiderBox = new JCheckBox(strings.getString("schneider"));
        schwarzBox = createSchwarzBox();

        final ActionListener actionListener = new ActionListener() {
            @Override
            public void actionPerformed(final ActionEvent e) {
                // FIXME (jan 28.11.2010) send sorting game type to JSkatMaster
                // --> more view components can benefit from this
                final GameType gameType = getSelectedGameType();

                if (gameType != null) {
                    userPanel.setSortGameType(gameType);

                    if (userPickedUpSkat) {
                        ouvertBox.setEnabled(gameType == GameType.NULL);
                    } else {
                        ouvertBox.setEnabled(true);
                        if (gameType != GameType.NULL) {
                            schneiderBox.setEnabled(true);
                            schwarzBox.setEnabled(true);
                            if (ouvertBox.isSelected()) {
                                schneiderBox.setSelected(true);
                                schwarzBox.setSelected(true);
                            }
                        } else {
                            schneiderBox.setEnabled(false);
                            schneiderBox.setSelected(false);
                            schwarzBox.setEnabled(false);
                            schwarzBox.setSelected(false);
                        }
                    }
                }
            }
        };

        gameTypeButtonGroup = new ButtonGroup();
        grandRadioButton = createRadioButton(GameType.GRAND, actionListener, gameTypeButtonGroup);
        clubsRadioButton = createRadioButton(GameType.CLUBS, actionListener, gameTypeButtonGroup);
        spadesRadioButton = createRadioButton(GameType.SPADES, actionListener, gameTypeButtonGroup);
        heartsRadioButton = createRadioButton(GameType.HEARTS, actionListener, gameTypeButtonGroup);
        diamondsRadioButton = createRadioButton(GameType.DIAMONDS, actionListener, gameTypeButtonGroup);
        nullRadioButton = createRadioButton(GameType.NULL, actionListener, gameTypeButtonGroup);

        panel.add(new JLabel(strings.getString("game")), "span 2, wrap");

        panel.add(clubsRadioButton);
        panel.add(spadesRadioButton, "wrap");
        panel.add(heartsRadioButton);
        panel.add(diamondsRadioButton, "wrap");
        panel.add(grandRadioButton);
        panel.add(nullRadioButton, "wrap");

        panel.add(new JLabel(strings.getString("win_levels")), "span 2, wrap");

        panel.add(handBox);
        panel.add(ouvertBox, "wrap");
        panel.add(schneiderBox);
        panel.add(schwarzBox, "wrap");

        final JButton announceButton = new JButton(actions.get(JSkatAction.ANNOUNCE_GAME));
        announceButton.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(final ActionEvent e) {

                if (getSelectedGameType() != null) {

                    try {
                        final GameAnnouncement announcement = getGameAnnouncement();
                        if (announcement == null) {
                            return;
                        }
                        e.setSource(announcement);
                    } catch (final IllegalArgumentException except) {
                        log.error(except.getMessage());
                    }
                } else {
                    e.setSource(null);
                }

                // fire event again
                announceButton.dispatchEvent(e);
            }

            private GameAnnouncement getGameAnnouncement() {

                final var gameType = getSelectedGameType();
                var contract = new GameContract(gameType);

                if (discardPanel.isUserLookedIntoSkat()) {

                    final CardList discardedCards = discardPanel.getDiscardedCards();

                    if (discardedCards.size() != 2) {

                        JSkatEventBus.INSTANCE.post(new InvalidNumberOfCardsInDiscardedSkatEvent());
                        return null;
                    }
                    if (GameType.NULL == gameType && ouvertBox.isSelected()) {
                        contract = contract.withOuvert(userPanel.cardPanel.getCards());
                    }

                    return new GameAnnouncement(contract, discardedCards);

                } else {

                    if (handBox.isSelected()) {
                        contract = contract.withHand();
                    }
                    if (schneiderBox.isSelected()) {
                        contract = contract.withSchneider();
                    }
                    if (schwarzBox.isSelected()) {
                        contract = contract.withSchwarz();
                    }
                    if (ouvertBox.isSelected()) {
                        contract = contract.withOuvert(userPanel.cardPanel.getCards());
                    }

                    return new GameAnnouncement(contract, CardList.empty());
                }
            }

        });
        panel.add(announceButton, "center, span 2");

        add(panel, "center");

        setOpaque(false);

        resetPanel();
    }

    private JRadioButton createRadioButton(final GameType gameType, final ActionListener actionListener, final ButtonGroup buttonGroup) {
        final JRadioButton radioButton = new JRadioButton(strings.getGameType(gameType));
        radioButton.setActionCommand(gameType.toString());
        radioButton.addActionListener(actionListener);
        buttonGroup.add(radioButton);
        return radioButton;
    }

    GameType getSelectedGameType() {
        for (final JRadioButton radioButton : new JRadioButton[]{grandRadioButton,
                spadesRadioButton,
                clubsRadioButton,
                diamondsRadioButton,
                heartsRadioButton,
                nullRadioButton}) {
            if (radioButton != null && radioButton.isSelected()) {
                return GameType.valueOf(radioButton.getActionCommand());
            }
        }
        return null;
    }

    private JCheckBox createOuvertBox() {
        final JCheckBox result = new JCheckBox(strings.getString("ouvert"));

        result.addItemListener(new ItemListener() {
            @Override
            public void itemStateChanged(final ItemEvent arg0) {
                if (result.isSelected() && handBox.isSelected()
                        && getSelectedGameType() != null) {
                    // hand ouvert
                    if (GameType.NULL != getSelectedGameType()) {
                        schneiderBox.setSelected(true);
                        schwarzBox.setSelected(true);
                    }
                }
            }
        });

        return result;
    }

    private JCheckBox createSchwarzBox() {
        final JCheckBox result = new JCheckBox(strings.getString("schwarz"));

        result.addItemListener(new ItemListener() {

            @Override
            public void itemStateChanged(final ItemEvent e) {
                if (result.isSelected()) {
                    schneiderBox.setSelected(true);
                }
            }
        });

        return result;
    }

    void resetPanel() {
        gameTypeButtonGroup.clearSelection();
        handBox.setSelected(true);
        ouvertBox.setSelected(false);
        schneiderBox.setSelected(false);
        schwarzBox.setSelected(false);
    }

    void setUserPickedUpSkat(final boolean isUserPickedUpSkat) {

        userPickedUpSkat = isUserPickedUpSkat;

        if (isUserPickedUpSkat) {
            handBox.setSelected(false);
            ouvertBox.setEnabled(GameType.NULL.equals(getSelectedGameType()));
            schneiderBox.setEnabled(false);
            schwarzBox.setEnabled(false);
        } else {
            handBox.setSelected(true);
            ouvertBox.setEnabled(true);
            schneiderBox.setEnabled(true);
            schwarzBox.setEnabled(true);
        }
    }
}

File: jskat-gui/src/main/kotlin/org/jskat/gui/javafx/table/GameAnnouncePanel.kt
package org.jskat.gui.javafx.table

import javafx.application.Platform
import javafx.geometry.Insets
import javafx.geometry.HPos
import javafx.geometry.Pos
import javafx.scene.control.*
import javafx.scene.layout.ColumnConstraints
import javafx.scene.layout.GridPane
import javafx.scene.layout.Priority
import javafx.scene.layout.Region
import org.jskat.control.JSkatEventBus
import org.jskat.control.event.skatgame.InvalidNumberOfCardsInDiscardedSkatEvent
import org.jskat.control.gui.action.JSkatAction
import org.jskat.control.gui.action.JSkatActionEvent
import org.jskat.data.GameAnnouncement
import org.jskat.data.GameContract
import org.jskat.gui.action.AbstractJSkatAction
import org.jskat.gui.img.JSkatGraphicRepository
import org.jskat.util.CardList
import org.jskat.util.GameType
import org.jskat.util.JSkatResourceBundle
import org.slf4j.LoggerFactory

class GameAnnouncePanel(
    private val actions: Map<JSkatAction, AbstractJSkatAction>,
    private val userPanel: JSkatUserPanel,
    private val discardPanel: DiscardPanel?
) : GridPane() {

    private val strings = JSkatResourceBundle.INSTANCE
    private val bitmaps = JSkatGraphicRepository.INSTANCE
    private val log = LoggerFactory.getLogger(GameAnnouncePanel::class.java)

    private val gameTypeGroup = ToggleGroup()
    private val grandButton = createRadioButton(GameType.GRAND)
    private val clubsButton = createRadioButton(GameType.CLUBS)
    private val spadesButton = createRadioButton(GameType.SPADES)
    private val heartsButton = createRadioButton(GameType.HEARTS)
    private val diamondsButton = createRadioButton(GameType.DIAMONDS)
    private val nullButton = createRadioButton(GameType.NULL)

    private val handBox = CheckBox(strings.getString("hand"))
    private val ouvertBox = CheckBox(strings.getString("ouvert"))
    private val schneiderBox = CheckBox(strings.getString("schneider"))
    private val schwarzBox = CheckBox(strings.getString("schwarz"))

    private val announceButton =
        Button(actions[JSkatAction.ANNOUNCE_GAME]?.getValue(AbstractJSkatAction.NAME) as? String ?: "")

    private var userPickedUpSkat = false

    init {
        initPanel()
    }

    private fun initPanel() {
        styleClass.add("action-panel")
        padding = Insets(8.0)
        hgap = 12.0
        vgap = 7.0
        alignment = Pos.CENTER

        columnConstraints.addAll(
            ColumnConstraints().apply { hgrow = Priority.ALWAYS },
            ColumnConstraints().apply { hgrow = Priority.ALWAYS }
        )

        val title = Label(announceButton.text).apply {
            styleClass.add("action-panel-title")
            maxWidth = Double.MAX_VALUE
        }
        add(title, 0, 0, 2, 1)

        val gameLabel = createSectionLabel(strings.getString("game"))
        add(gameLabel, 0, 1, 2, 1)
        add(clubsButton, 0, 2)
        add(spadesButton, 1, 2)
        add(heartsButton, 0, 3)
        add(diamondsButton, 1, 3)
        add(grandButton, 0, 4)
        add(nullButton, 1, 4)

        val levelsLabel = createSectionLabel(strings.getString("win_levels"))
        add(levelsLabel, 0, 5, 2, 1)
        add(handBox, 0, 6)
        add(ouvertBox, 1, 6)
        add(schneiderBox, 0, 7)
        add(schwarzBox, 1, 7)

        announceButton.graphic =
            bitmaps.getImageView(JSkatGraphicRepository.Icon.PLAY, JSkatGraphicRepository.IconSize.BIG)
        announceButton.maxWidth = Region.USE_PREF_SIZE
        add(announceButton, 0, 8, 2, 1)
        GridPane.setHalignment(announceButton, HPos.CENTER)

        handBox.isDisable = true

        gameTypeGroup.selectedToggleProperty().addListener { _, _, newToggle ->
            if (newToggle != null) {
                val gameType = (newToggle as RadioButton).userData as GameType
                updateGameType(gameType)
            }
        }

        ouvertBox.selectedProperty().addListener { _, _, isSelected ->
            if (isSelected && handBox.isSelected && getSelectedGameType() != null) {
                if (GameType.NULL != getSelectedGameType()) {
                    schneiderBox.isSelected = true
                    schwarzBox.isSelected = true
                }
            }
        }

        schwarzBox.selectedProperty().addListener { _, _, isSelected ->
            if (isSelected) {
                schneiderBox.isSelected = true
            }
        }

        announceButton.setOnAction {
            announceGame()
        }

        resetPanel()
    }

    private fun createRadioButton(gameType: GameType): RadioButton {
        val button = RadioButton(strings.getGameType(gameType))
        button.userData = gameType
        button.toggleGroup = gameTypeGroup
        return button
    }

    private fun createSectionLabel(text: String): Label {
        return Label(text).apply {
            styleClass.add("action-panel-section-label")
        }
    }

    private fun updateGameType(gameType: GameType) {
        Platform.runLater {
            userPanel.setSortGameType(gameType)
        }

        if (userPickedUpSkat) {
            ouvertBox.isDisable = (gameType != GameType.NULL)
        } else {
            ouvertBox.isDisable = false
            if (gameType != GameType.NULL) {
                schneiderBox.isDisable = false
                schwarzBox.isDisable = false
                if (ouvertBox.isSelected) {
                    schneiderBox.isSelected = true
                    schwarzBox.isSelected = true
                }
            } else {
                schneiderBox.isDisable = true
                schneiderBox.isSelected = false
                schwarzBox.isDisable = true
                schwarzBox.isSelected = false
            }
        }
    }

    private fun getSelectedGameType(): GameType? {
        return gameTypeGroup.selectedToggle?.userData as? GameType
    }

    private fun announceGame() {
        val gameType = getSelectedGameType() ?: return

        val isHand = handBox.isSelected
        val isOuvert = ouvertBox.isSelected
        val isSchneider = schneiderBox.isSelected
        val isSchwarz = schwarzBox.isSelected
        val pickedUpSkat = discardPanel?.userPickedUpSkat ?: false
        val discardedCards = discardPanel?.discardedCards ?: CardList()
        val userHand = CardList(userPanel.getHandCards())

        try {
            var contract = GameContract(gameType)

            if (pickedUpSkat) {
                if (discardedCards.size() != 2) {
                    JSkatEventBus.INSTANCE.post(InvalidNumberOfCardsInDiscardedSkatEvent())
                    return
                }

                if (GameType.NULL == gameType && isOuvert) {
                    contract = contract.withOuvert(userHand)
                }

                val announcement = GameAnnouncement(contract, discardedCards)
                fireAnnounceAction(announcement)

            } else {
                if (isHand) contract = contract.withHand()
                if (isSchneider) contract = contract.withSchneider()
                if (isSchwarz) contract = contract.withSchwarz()
                if (isOuvert) {
                    contract = contract.withOuvert(userHand)
                }

                val announcement = GameAnnouncement(contract, discardedCards)
                fireAnnounceAction(announcement)
            }
        } catch (e: Exception) {
            log.error(e.message)
        }
    }

    private fun fireAnnounceAction(announcement: GameAnnouncement) {
        actions[JSkatAction.ANNOUNCE_GAME]?.actionPerformed(JSkatActionEvent(JSkatAction.ANNOUNCE_GAME, announcement))
    }

    fun resetPanel() {
        Platform.runLater {
            gameTypeGroup.selectToggle(null)
            handBox.isSelected = true
            ouvertBox.isSelected = false
            schneiderBox.isSelected = false
            schwarzBox.isSelected = false
        }
    }

    fun setUserPickedUpSkat(isUserPickedUpSkat: Boolean) {
        Platform.runLater {
            userPickedUpSkat = isUserPickedUpSkat
            if (isUserPickedUpSkat) {
                handBox.isSelected = false
                ouvertBox.isDisable = (GameType.NULL == getSelectedGameType())
                schneiderBox.isDisable = true
                schwarzBox.isDisable = true
            } else {
                handBox.isSelected = true
                ouvertBox.isDisable = false
                schneiderBox.isDisable = true
                schwarzBox.isDisable = true
            }
        }
    }
}
